### ISSUE TITLE
Refactor `struct` types to an annotation `@[Struct]`

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#13017]
For Crystal 2.0, I think one idea could be to have everything be a `class`. If you want something to have struct semantics you can use an annotation, like `@[Struct]`. That way when you reopen a type, it's always using `class` and there's less chance for a breaking change.

Of course there's still `module`, but changing things from class to module is less common.

_Originally posted by @asterite in https://github.com/crystal-lang/crystal/issues/13015#issuecomment-1406443477_

**Comment by yxhuvud** [comment]
So the argument is to remove the struct keyword to .. make it easier to reopen stdlib base types? 

If the purpose is to make it easier to reopen types, could we instead get a construct that explicitly open an *existing* type (of any type) and fail if it doesn't already exist? Then accidental conflicts would get caught and the risk of defining stuff in the wrong order would be less. The ability to reopen classes are great, but it should be a rare occurrence, and locking it down to be more explicit with regards to what is redefined or created from scratch would probably be an improvement in any case.

/not a big fan of annotations

**Comment by asterite** [comment]
Right. I proposed a reopen keyword in the past but it didn't seem to be liked.

**Comment by BlobCodes** [comment]
I think using an annotation would look messy.
Also, currently annotations are not inherited. Because of that, I think it doesn't make sense to assume the `@[Struct]` annotation would be inherited too.

Instead, I think this could be done by requiring types to explicitly inherit from `Struct` and drop the `struct` keyword:
```crystal
class Foo < Struct
end
```

This way, types implicitly inherit from `Reference` and can explicitly specify a type if that's not wanted.

I think this would also make sense because currently, numeric types which inherit from `Number` do ***not*** inherit from `Struct`, even though they are defined using the `struct` keyword. This all currently seems like a bodge.
By using this proposal, that would actually make sense because `Number` is another type inheriting from `Value`, not `Struct`.

This would work because crystal does not require explicitly specifying the parent type when reopening a class:
```crystal
class A
end

class B < A
end

# Reopen without specifying parent (this is currently valid code)
class B
  def hi
    1
  end
end

# Prints "1" using `crystal eval`
puts B.new.hi
```

---

As an additional benefit, this also makes the type system a bit more flexible. For example, in the future a type `Local` could be created for stack-allocated pass-by-reference types, which directly inherits from `Object`.
